feat: Glob patterns for `cacheTtlOverride` keys

zharinov opened the pull request:
<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->
<!-- Please read https://github.com/renovatebot/renovate/blob/main/.github/contributing.md before you create your pull request.-->

## Changes

<!-- Describe what behavior is changed by this PR. -->

## Context

- Closes: #36836
<!-- Describe why you're making these changes if it's not already explained in a corresponding issue. -->
<!-- If you're closing an existing issue with this pull request, use the keyword Closes #issue_number. -->
<!-- If you're referencing an issue with this pull request, put it in a Markdown list like this: - #issue_number. -->

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

zharinov committed:
feat: Glob patterns for `cacheTtlOverride` keys

rarkins reviewed (changes requested):
Needs docs update

zharinov committed:
Update self-hosted-configuration.md

zharinov committed:
Update self-hosted-configuration.md

zharinov committed:
Update self-hosted-configuration.md

zharinov committed:
Update self-hosted-configuration.md

zharinov committed:
Simplify code

zharinov committed:
Merge branch 'main' into feat/ttl-override-wildcard

zharinov committed:
Merge branch 'main' into feat/ttl-override-wildcard

rarkins commented on docs/usage/self-hosted-configuration.md:
```diff
@@ -295,20 +295,46 @@ In the self-hosted setup, use option to enable caching of private packages to im
+- `datasource-docker-tags` gets 120 minutes (exact match)
+- `datasource-npm` gets 90 minutes (exact match)
+- Other datasources like `datasource-maven` get 15 minutes (matches `*` which is evaluated last)
+- Changelog namespaces like `changelog-github` get 30 minutes (regex match)
```
Why is a regex match higher priority? Also, you didn't mention it in the "Priority order" earlier

zharinov replied:
Let's do the "first match wins" because I think it's the behavior people expect by default.

zharinov committed:
Update cache TTL pattern matching logic

The cache TTL override system now evaluates patterns in order of
appearance rather than using last-match-wins. This better matches user
expectations and makes the behavior more predictable.

Pattern evaluation stops at the first valid match found, respecting the
order specified in the configuration object. Test cases have been
updated to reflect this change.

rarkins commented on docs/usage/self-hosted-configuration.md:
```diff
@@ -295,20 +295,46 @@ In the self-hosted setup, use option to enable caching of private packages to im
+Priority order:
+
+1. Exact namespace matches take highest priority
+2. If no exact match, patterns are evaluated in the order they appear in the object, and the first matching pattern wins
```
Should be longest match first

zharinov committed:
Use longest pattern for cache TTL overrides

zharinov committed:
Fix doc

zharinov committed:
Update doc

zharinov commented on docs/usage/self-hosted-configuration.md:
```diff
@@ -367,14 +396,15 @@ Valid codes for namespaces are as follows:
 - `datasource-maven:postprocess-reject`
 - `datasource-node-version`
-- `datasource-npm:data`
+- `datasource-npm:cache-provider`
```
Yes, the list was a bit outdated

zharinov replied:
Maybe we need to point at the source code instead

rarkins replied:
No, it needs to be docs. Maybe we need a test to ensure it matches. When did this change happen?

zharinov replied:
https://github.com/renovatebot/renovate/pull/36909

zharinov commented on lib/util/cache/package/ttl.spec.ts:
```diff
@@ -0,0 +1,380 @@
+            'datasource-npm': 'invalid',
+          },
+        });
+        const result = resolveTtlValues('datasource-npm' as never, 60);
```
Not assignable, I just wanted to use whatever strings because it doesn't important to use only existing namespaces here

zharinov committed:
Merge branch 'main' into feat/ttl-override-wildcard

zharinov committed:
Merge branch 'main' into feat/ttl-override-wildcard
